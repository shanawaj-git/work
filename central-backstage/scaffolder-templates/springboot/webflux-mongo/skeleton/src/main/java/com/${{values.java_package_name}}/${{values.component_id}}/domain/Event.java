package com.${{values.java_package_name}}.${{values.component_id}}.domain;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@AllArgsConstructor
@NoArgsConstructor
@Data
public class Event<T> {
    protected String topic;
    protected String eventType;
    protected T data;
}
