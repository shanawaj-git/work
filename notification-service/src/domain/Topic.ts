enum Topic {
  Prescription = 'prescriptions',
  Authentication = 'authentication',
}

export enum EventType {
  // prescriptions
  PresciptionReceived = 'prescriptions.received',

  // autehntication
  OtpGenerated = 'authentication.otpGenerated',
}

export const EVENT_TYPE_PER_TOPIC = {
  [Topic.Prescription]: [EventType.PresciptionReceived],
  [Topic.Authentication]: [EventType.OtpGenerated],
};

export default Topic;
