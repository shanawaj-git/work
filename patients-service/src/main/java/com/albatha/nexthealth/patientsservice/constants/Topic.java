package com.albatha.nexthealth.patientsservice.constants;

public enum Topic {

    PATIENTS("patients");
    public final String label;
    Topic(String label) {
        this.label = label;
    }

}
