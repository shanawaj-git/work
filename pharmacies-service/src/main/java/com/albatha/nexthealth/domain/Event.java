package com.albatha.nexthealth.domain;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@AllArgsConstructor
@NoArgsConstructor
@Data
public class Event<T> {
    protected String topic;
    protected String eventType;
    protected T data;
}
