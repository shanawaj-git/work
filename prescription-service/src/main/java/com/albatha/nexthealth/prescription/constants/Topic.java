package com.albatha.nexthealth.prescription.constants;

public enum Topic {
    PRESCRIPTIONS("prescriptions");
    public final String name;
    private Topic(String name) {
        this.name = name;
    }
}
