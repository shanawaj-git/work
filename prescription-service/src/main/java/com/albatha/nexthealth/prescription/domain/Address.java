package com.albatha.nexthealth.prescription.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import java.sql.Timestamp;

@Builder
@NoArgsConstructor
@AllArgsConstructor
@Data
@Entity
public class Address {

    @Id
    @SequenceGenerator(name = "address_sequence", sequenceName = "address_sequence", allocationSize = 1)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "address_sequence")
    private Long addressId;
    private String addressLine1;
    private String addressLine2;
    private String area;
    private String city;
    private String state;
    private String postalCode;
    private String type;
    private String country;

    @CreationTimestamp
    private Timestamp createdDate;
}
