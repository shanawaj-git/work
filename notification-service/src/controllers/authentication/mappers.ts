import Topic, { EventType } from 'src/domain/Topic';
import { SendSMSInput } from 'src/sms/sms.interface';
import { ERROR_CODES } from 'src/common/constants';
const ERRORS = {
  MISSING_TEMPLATE_CODE: {
    code: ERROR_CODES.MISSING_TEMPLATE_CODE,
    message: 'missing template code',
  },
};

export enum TemplateCodes {
  OTP_GENERATED = 'opt_generated',
}

type OtpData = {
  otp: string;
  phoneNumber: string;
};

const TEMPLATE_CODE_BY_EVENT_TYPE = {
  [EventType.OtpGenerated]: TemplateCodes.OTP_GENERATED,
};

export const getTemplateParamsByTemplateCode = (
  otpData: OtpData,
  templateCode: string,
): Record<string, string> => {
  switch (templateCode) {
    case TemplateCodes.OTP_GENERATED: {
      const { otp } = otpData;

      return {
        otp,
      };
    }
    default: {
      throw new Error(ERRORS.MISSING_TEMPLATE_CODE.message);
    }
  }
};

export const mapOtpDataToSmsInput = ({ otpData, eventType }): SendSMSInput => {
  const { phoneNumber } = otpData;

  return {
    templateCode: TEMPLATE_CODE_BY_EVENT_TYPE[eventType],
    recipient: phoneNumber,
    params: getTemplateParamsByTemplateCode(
      otpData,
      TEMPLATE_CODE_BY_EVENT_TYPE[eventType],
    ),
    topic: Topic.Authentication,
    referenceNumberInTopic: phoneNumber,
  };
};
