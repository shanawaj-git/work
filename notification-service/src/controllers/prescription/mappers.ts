import Topic, { EventType } from 'src/domain/Topic';
import { Environment } from 'src/common/constants';
import { SendSMSInput } from 'src/sms/sms.interface';
import { ERROR_CODES } from 'src/common/constants';
const ERRORS = {
  MISSING_TEMPLATE_CODE: {
    code: ERROR_CODES.MISSING_TEMPLATE_CODE,
    message: 'missing template code',
  },
};

const prescriptionsBaseUrl = {
  [Environment.PRODUCTION]: '',
  [Environment.DEVELOPMENT]: 'http://localhost:3000',
};

const PARAM_REDIRECT = 'redirect';
const CONTEXT_PATH_PRESCRIPTIONS = '/prescriptions';

export enum TemplateCodes {
  NEW_PRESECRIPTION = 'new_prescription',
}

type Prescription = {
  patient: {
    name: {
      first: string;
      middle: string;
      last: string;
    };
    mobile;
  };
  prescriptionNumber: string;
};

const TEMPLATE_CODE_BY_EVENT_TYPE = {
  [EventType.PresciptionReceived]: TemplateCodes.NEW_PRESECRIPTION,
};

export const getTemplateParamsByTemplateCode = (
  prescription: Prescription,
  templateCode: string,
): Record<string, string> => {
  switch (templateCode) {
    case TemplateCodes.NEW_PRESECRIPTION: {
      const { NODE_ENV } = process.env;

      const { patient, prescriptionNumber } = prescription;
      const { name } = patient;
      const prescriptionLink = new URL(
        prescriptionsBaseUrl[NODE_ENV] ||
          prescriptionsBaseUrl[Environment.DEVELOPMENT],
      );
      prescriptionLink.searchParams.append(
        PARAM_REDIRECT,
        `${CONTEXT_PATH_PRESCRIPTIONS}/${prescriptionNumber}`,
      );
      return {
        patientName: `${name.first} ${name.last}`,
        prescriptionUrl: prescriptionLink.href,
      };
    }
    default: {
      throw new Error(ERRORS.MISSING_TEMPLATE_CODE.message);
    }
  }
};

export const mapPrescriptionDataToSmsInput = ({
  prescription,
  eventType,
}): SendSMSInput => {
  const { patient, prescriptionNumber } = prescription;
  const { mobileNumber } = patient;

  return {
    templateCode: TEMPLATE_CODE_BY_EVENT_TYPE[eventType],
    recipient: mobileNumber,
    params: getTemplateParamsByTemplateCode(
      prescription,
      TEMPLATE_CODE_BY_EVENT_TYPE[eventType],
    ),
    topic: Topic.Prescription,
    referenceNumberInTopic: prescriptionNumber,
  };
};
