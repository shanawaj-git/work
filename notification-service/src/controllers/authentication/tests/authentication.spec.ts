import { mapOtpDataToSmsInput } from '../mappers';
import { EventType } from 'src/domain/Topic';
describe('mapOtpDataToSmsInput', () => {
  const input = {
    otpData: {
      otp: '233344',
      phoneNumber: '+971526234727',
    },
    eventType: EventType.OtpGenerated,
  };

  it('should map otp data to sms input', () => {
    expect(mapOtpDataToSmsInput(input)).toMatchSnapshot();
  });

  it('should throw in case of undefined event type', () => {
    const missingEventTypeInput = {
      ...input,
      eventType: undefined,
    };

    expect(() => mapOtpDataToSmsInput(missingEventTypeInput)).toThrow(
      'missing template code',
    );
  });
});
