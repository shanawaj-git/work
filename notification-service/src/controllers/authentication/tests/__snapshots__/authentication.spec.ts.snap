// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapOtpDataToSmsInput should map otp data to sms input 1`] = `
Object {
  "params": Object {
    "otp": "233344",
  },
  "recipient": "+971526234727",
  "referenceNumberInTopic": "+971526234727",
  "templateCode": "opt_generated",
  "topic": "authentication",
}
`;
